illumos-utils/src/opte/port_manager.rs

```diff
@@ -1041,37 +952,432 @@ fn system_routes_v4<'a>(
     routes: &'a mut HashMap<RouterId, RouteSet>,
     port: &Port,
 ) -> &'a mut RouteSet {
-    let routes = routes.entry(port.system_router_key()).or_insert_with(|| {
-        let routes = HashSet::new();
-        RouteSet { version: None, routes, active_ports: 0 }
-    });
-
-    if is_service {
-        routes.routes.insert(ResolvedVpcRoute {
-            dest: "0.0.0.0/0".parse().unwrap(),
-            target: ApiRouterTarget::InternetGateway(None),
-        });
-    }
-
-    routes
+    let maybe_default_dest = if is_service {
+        Some(IpNet::V4(Ipv4Net::new(Ipv4Addr::UNSPECIFIED, 0).unwrap()))
+    } else {
+        None
+    };
+    fetch_or_insert_system_routes(routes, port, maybe_default_dest)
 }
 
 fn system_routes_v6<'a>(
     is_service: bool,
     routes: &'a mut HashMap<RouterId, RouteSet>,
     port: &Port,
 ) -> &'a mut RouteSet {
-    let routes = routes.entry(port.system_router_key()).or_insert_with(|| {
-        let routes = HashSet::new();
-        RouteSet { version: None, routes, active_ports: 0 }
-    });
+    let maybe_default_dest = if is_service {
+        Some(IpNet::V6(Ipv6Net::new(Ipv6Addr::UNSPECIFIED, 0).unwrap()))
+    } else {
+        None
+    };
+    fetch_or_insert_system_routes(routes, port, maybe_default_dest)
+}
 
-    if is_service {
+fn fetch_or_insert_system_routes<'a>(
+    routes: &'a mut HashMap<RouterId, RouteSet>,
+    port: &Port,
+    maybe_default_dest: Option<IpNet>,
+) -> &'a mut RouteSet {
+    let routes = routes.entry(port.system_router_key()).or_default();
+    if let Some(dest) = maybe_default_dest {
         routes.routes.insert(ResolvedVpcRoute {
-            dest: "::/0".parse().unwrap(),
-            target: ApiRouterTarget::InternetGateway(None),
+            dest,
+            // Always insert a rule targeting the _system VPC Internet Gateway_.
+            // This may be sent later from Nexus, but we need it during
+            // bootstrapping NTP or other very early services, before the
+            // control plane database has been started.
+            target: ApiRouterTarget::InternetGateway(
+                InternetGatewayRouterTarget::System,
```

Comment:
This feels a little iffy to me. We're taking an `Option<IpNet>` and assuming that if it's `Some(_)`, that means we should add a route for that under the `System` internet gateway target. A couple ideas for alternatives:

* Go back to taking a `bool` here; then we get decide both the dest and the router target based on that boolean (so they're obviously consistent with each other)
* Take an `Option<ResolvedVpcRoute>` instead, so our caller is responsible for making sure the dest and router target are consistent

Reply:
I switched back to the two separate methods in 2024b245c. It wasn't worth the complexity to try to de-dupe that.